### ISSUE TITLE
"Connection timed out" exception

### ISSUE THREAD

**Issue by ilya-levin-lokalise** [phpredis/phpredis#2085]
Hi everyone! Couldn't find a better place to ask my question.
We have a high-load project that utilizes Redis a lot. The problem is - we occasionally are having `RedisException` being thrown with `Connection timed out` message, even though the timeout is pretty high (several seconds). We're trying to track down the issue and our monitoring data shows that there are no issues with connection from PHP host to Redis host, however, the PHP application itself is occasionally throws this exception.
Long story short - I've downloaded the source of the module (this repo) and was trying to find `Connection timed out` string or the place where this exception is thrown and failed. I wanted to check what are the conditions for this exception to be thrown. Can someone point me out?

**Comment by Venorcis** [comment]
We actually have the same issue since about a month now (I'd say since the 5.3.7 release), just random Connection Timeouts happening, no luck in tracking anything down in our servers (we even doubled the capacity of all servers to try and resolve it).

Edit: I actually just checked, and we updated to phpredis 5.3.7 on the 28th of February (sourced from the remi rpm), and right from that date on these timeouts started happening. We'll try downgrading to 5.3.6 to see if that helps.

**Comment by ilya-levin-lokalise** [comment]
Our project is currently using a pretty old 4.3.0 version of the extension and facing the same issue.

**Comment by Venorcis** [comment]
> We actually have the same issue since about a month now (I'd say since the 5.3.7 release), just random Connection Timeouts happening, no luck in tracking anything down in our servers (we even doubled the capacity of all servers to try and resolve it).
> 
> Edit: I actually just checked, and we updated to phpredis 5.3.7 on the 28th of February (sourced from the remi rpm), and right from that date on these timeouts started happening. We'll try downgrading to 5.3.6 to see if that helps.

Actually the downgrade does not help. However we did also upgrade from PHP 8.0 to 8.1 at that same day, so now suspecting that that's related.

**Comment by deniscsz** [comment]
I have experiencing the same scenario with PHP 8.1.

**Comment by fwilliamconceicao** [comment]
I'm having this issue with PHP 8.2.

**Comment by chillidoor** [comment]
We're seeing this issue with PHP 8.1 and 7.3

**Comment by dellow** [comment]
Also seeing this with 8.2

**Comment by pablorsk** [comment]
Hello, maybe our context can be helpful.

**Only happens in docker (php:8.1-fpm-alpine) when PHP is running under crond**. We precisely use Laravel and it happens with `php artisan schedule:run`.

We have exactly the same problem under the same circumstances:
- The error is random.
- There is no lack of resources nor are there any peaks in the history.
- The timeout can reach up to 2 minutes.
- Even a simple and random redis get() fails

We are currently in dialogue with DigitalOcean support, since we use the database managed by them.

**Comment by deniscsz** [comment]
What worked to me: make the tls version explicit in the database connection.

For some reason when I tried use `tlsv1.2` in the Redis connection in `scheme` parameter the issue is gone.

I found this "solution" in this blog post: https://alexvanderbist.com/2021/slow-queues-and-redis-over-tls/

**Comment by chillidoor** [comment]
In our case, we were already using tls 1.2. 

Similar to @pablorsk, we're using the official PHP docker images, and also are using DigitalOcaen's managed Redis database. Increasing the time to >120s hasn't solve the problem for us.

**Comment by fwilliamconceicao** [comment]
> In our case, we were already using tls 1.2.
> 
> Similar to @pablorsk, we're using the official PHP docker images, and also are using DigitalOcaen's managed Redis database. Increasing the time to >120s hasn't solve the problem for us.

This doesn't have a relation with TLS 1.2, we've been using it here for 2 years and this issue started triggering randomly in April until now, but now it doesn't happen with frequency. 

But at least once per week (on a random day) that happens.

**Comment by pablorsk** [comment]
We have the same case to @fwilliamconceicao, no related with tls. We try with persistent=1, read_write_timeout=0 and read_write_timeout=-1.

Maybe this information will be helpful. This error only happens when the php is executed by crond. In detail, the structure is:
Docker php:8.1-fpm -> supervisord -> crond -> `php artisan schedule:run`

This PHP fails randomly, between 30 and 180 minutes, on any REDIS.GET. The script does a get and freezes, finally causing a timeout.

**Comment by pablorsk** [comment]
UPDATE
- Not related with supervisord or crond, also fails randomly without those tools.
- Not related with CPU/Memory resources. I have upgraded the machine to one with two dedicated cores and the problem persists.

I just have to try implementing my own redis on our k8s cluster.

**Comment by fwilliamconceicao** [comment]
> UPDATE
> 
> * Not related with supervisord or crond, also fails randomly without those tools.
> * Not related with CPU/Memory resources. I have upgraded the machine to one with two dedicated cores and the problem persists.
> 
> I just have to try implementing my own redis on our k8s cluster.

For sure it's not related to any tool, I tried to remove any framework behind it (here we use Laravel + Horizon) and the problem persisted even with a fresh install, I tried with a machine with 4vCPU (dedicated as well) and the problem persist.

It's something related to how the library uses the native PHP resources to communicate with REDIS, anyway, this issue seems to me like it'll be abandoned. 

might be in the future I'll fork the library and fix it (like I did with other projects) and keep going.

**Comment by cyppe** [comment]
Nothing to add more than that I am also seeing the same issue.  Only for php artisan schedule:run, and only when it runs in cron.

Works fine when I run command manually.

**Comment by YonkoSam** [comment]
this is still happening as of 6-6-2025

**Comment by oussamaHer** [comment]
Hi,
Having the same Redis extension issue on both php:8.3-cli and php:8.3-cli-composer Docker images .
The setup involves multiple crons running inside Kubernetes.

**Comment by rajmundtoth0** [comment]
Same here, without cron or anything. Just happens randomly

**Comment by jarnovanleeuwen** [comment]
In our case it looks like to be a firewall issue on the Redis server.

We are using DigitalOcean Manages Kubernetes and we sometimes get this exception when connecting to a Managed Redis/Valkey database after starting a new pod on a recently created node, e.g. after Horizontal Pod Autoscaling. When the new pod/node has been running for a few minutes, PhpRedis can connect and the issue never reappears.

We are using trusted sources on the Managed Redis/Valkey database and our guess is that it takes some time for new IP addresses to be allowed after creating a new Kubernetes node. We have now added the IP CIDR ranges (e.g. `10.133.0.0/16`) of the Kubernetes' VPC network to the trusted sources (Node network, Pod network and Service network).

The problem seems to be less frequent, but is still not fully gone. Because the problems only appears when starting a new pod, you can make sure that the pod can connect to Redis before serving traffic using a Startup Probe or Readiness Probe. Once the probe succeeds, the problem doesn't reappear for that pod.
